docs: Add the Release Signal Handbook (#2352)
kubernetes/sig-release: release-team/role-handbooks/release-signal/README.md

# Release Signal Role Handbook

<!-- TOC -->
* [Release Signal Role Handbook](#release-signal-role-handbook)
  * [Overview](#overview)
  * [Requirements](#requirements)
    * [Time Requirements](#time-requirements)
    * [Additional Requirements for Shadows](#additional-requirements-for-shadows)
    * [Additional Requirements for Leads](#additional-requirements-for-leads)
  * [Task Overview](#task-overview)

>     * [How to Escalate](#how-to-escalate)

Author: Done.

  * [Judging State of Testgrid dashboards](#judging-state-of-testgrid-dashboards)
  * [Blocking vs. Informing Dashboards](#blocking-vs-informing-dashboards)
  * [Opening Issues](#opening-issues)
    * [Decision Tree](#decision-tree)
    * [Closing Issues](#closing-issues)
  * [Special high risk test categories to monitor](#special-high-risk-test-categories-to-monitor)
    * [Scalability tests](#scalability-tests)
  * [Reporting](#reporting)
  * [Tips and Tricks of the game](#tips-and-tricks-of-the-game)
    * [A Tour of CI on the Kubernetes Project](#a-tour-of-ci-on-the-kubernetes-project)
    * [Checking test dashboards](#checking-test-dashboards)
    * [Priority Labels](#priority-labels)
    * [Milestones](#milestones)
<!-- TOC -->

## Overview

The release signal team is responsible for making sure that **both** Issues (`is:issue`) and Pull Requests (PRs) (`is:pr`), which
are targeted for the ongoing release cycle, are dealt with in a timely fashion. It also assumes the responsibility of the quality
gate for the release, checking the health status of testgrid dashboards. This team is a combination of old CI Signal and Bug

> Triage sub-team. Further responsibilities are:

Author: Yeah, there was some repetition, slimmed the list.

> Triage sub-team. Further responsibilities are:

Reviewer: > This team is a combination of old CI Signal and Bug
Triage sub-team.

Adding a link to these would be nice: [CI Signal](https://github.com/kubernetes/sig-release/tree/master/release-team/role-handbooks/ci-signal) and [Bug
Triage](https://github.com/kubernetes/sig-release/tree/master/release-team/role-handbooks/bug-triage).

> - Continuously monitoring various e2e tests in sig-release

Author: Done, combined the two points.

  dashboards ([master-blocking](https://testgrid.k8s.io/sig-release-master-blocking), [master-informing](https://testgrid.k8s.io/sig-release-master-informing), `release-x.y-blocking/informing` (
  x.y being the current release)) throughout the release cycle

> - Providing early and ongoing signals on release and test health to both the Release team and various SIGs

Author: Done.

> - Ensuring that all release-blocking tests provide a clear Go/No-Go signal for the release

Reviewer 2: ```suggestion
Give a “Go” or “No-Go” Signal to the release team whenever a release is being cut.
```

Author: Done.

> - Flagging regressions as close to the source as possible, i.e., as soon as the offending code was merged

Reviewer 2: Implicitly mentioned

Author: Done, slimmed the list.

- Filing issues proactively for test failures and flakes, triaging issues to appropriate SIG/owner, following up on status and
  verifying fixes on newer test runs
    - See also: [How to escalate issues](#how-to-escalate)
- Studying patterns/history of test health from previous releases and working closely with SIGs and test owners to
    - Understand and fix the test for the current release
    - Understand if the test needs to be on the release blocking dashboard

>     - Work with SIG-Testing on any possible test infra improvement to help improve the test pass rate

Reviewer 2: Studying patterns and history is not really an actionable responsibility. 

"Work with SIG-Testing on any possible test infra improvement to help improve the test pass rate" This goes in an actionable direction.

Maybe we could cut this?

Author: Yeah, we can cut this. If a test fails due to infra, it is just like a failure for any other SIG, we would like to get the issue resolved, whatever the reason may be. 

Removed it, done.

- Making recommendations to SIG-Release for promoting and demoting release blocking and merge blocking tests as per
  the [Blocking Jobs Criteria](/release-blocking-jobs.md)

> - Listing open issues and PRs targeted for the ongoing release cycle

Reviewer 2: reporting issues and PRs right? we already mentioned this

Author: Done, slimmed the list.

> - Communicating with relevant assignees, owners, and SIG leads of issues/PRs to get the status

Reviewer 2: coordinate with sigs already mentioned

Author: Done, slimmed the list.

- Updating issues/PRs, clarifying situations, enabling next level decision making

> - Presenting summary reports at release team and burndown meetings

Author: Done, slimmed the list.

- Maintaining the automation around tracking issues/PRs against the current milestone

The core responsibility of the Release Signal lead is to foster a culture of continuous test integration, fail-fast and fix-fast
mindset and strive for continuously passing tests that provide an accurate positive signal to the release team to ensure that
releases go out on time with high quality. It is critical to maintain a sustained focus on test health through the entire release
cycle, as opposed to accepting destabilizing test failures during active enhancement development followed by an extended test

> stabilization phase.

Author: Done, moved it to leads section.

## Requirements

### Time Requirements

Release Signal is one of the most time-intensive roles on the release team. As such, it is not recommended for candidates who do

> not have employer support for their work on the Release Team. General time requirements:

Author: It was in the previous role handbook, but yeah, it can be totally done without employer support.

Removed it, done.

- 1/2 hour to 2 hours daily, every workday, checking and following up on failing/flaky tests, issues and PRs targeted for the

>   ongoing release cycle around week 6 (mid-release).

Reviewer 2: It is also possible to rotate test & issue monitoring across the team, so a shadow may only need to work 1h on Monday or so (+ attend meetings).

>   ongoing release cycle around week 6 (mid-release).

Author: The reason why I mentioned daily for all shadows is that, sometimes, some shadows are not active, even when they are the prime observer. So this is sort of a safeguard, I updated it to 10 minutes daily.

Done, separated sections and decreased time commitment for shadows.

- Availability to attend most of Release Team (weekly) and Burndown meetings (daily during Code Freeze), subject to time zone
  appropriateness.
- Ability to follow up on test fixes during Code Freeze at arbitrary times to ensure rapid turnaround.
- The time commitment increases through the release cycle, peaking during Code Freeze. In the last week of Code Freeze, shadows
  should expect to spend at least 10 hours and lead at least 20 hours per week.

### Additional Requirements for Shadows

The following requirements are all things that the lead should already have, and shadows should acquire, with coaching, within the
first weeks of the release cycle:

- Have signed the contributor CLA for Kubernetes.
- Become a Kubernetes org member. In many cases, this will be done with the sponsorship of the Release Signal Lead or Release Lead
  in the first week of the cycle.
    - The process to become an org member is

>       in [our community membership ladder](https://github.com/kubernetes/community/blob/master/community-membership.md#requirements-for-outside-collaborators)

Reviewer 2: can we link to another document. These are general requirements across the team, which we don't want to repeat in every handbook.

Author: Linked to: https://github.com/kubernetes/sig-release/blob/master/release-team/release-team-onboarding.md#kubernetes-community-membership

Done.

- General familiarity with our [test boards](https://testgrid.k8s.io/) and examining test results from automated tests.
- Willingness and ability to follow up with contributors about test failures on Slack, email, Discuss, and SIG meetings, as

>   appropriate.

Reviewer 2: Do we really use discuss? :D

Author: Again, copied from the previous handbook, slack is sufficient IMO.

Removed, it done.

- Ability to file well-crafted Kubernetes issues, including labelling.
- General familiarity with GitHub labels and how to find issues/PRs for the current milestone.
- General knowledge of our SIG system and SIGs' areas of responsibility.

> Additionally, the following qualifications make a candidate more suitable for the Release Signal team but are not requirements:

Reviewer 2: I like this section. Can we highlight it so it stands out a bit more as "beneficial experience" or similar

Author: Done, highlighted it as a section.

> - Prior involvement with SIG Testing and the Test Infrastructure team.

Reviewer 2: missing link to sig-testing & infra

Author: Done, added.

- Experience with automated testing, CI/CD, quality engineering, and/or quality assurance.
- Experience with quality assurance and bug-tracking systems.
- Experience with the GitHub automation.

### Additional Requirements for Leads

In addition to the above requirements for shadows, most of which become prerequisites, Release Signal Leads must:

- Have the ability to add a milestone to issues, so must be a member of

>   the [milestone maintainers](https://github.com/orgs/kubernetes/teams/milestone-maintainers).

Reviewer 2: as far as i remember the release lead updates the team at the start of the cycle. We could mention this here.

Author: Yes, Release Lead does that, added. Done.

> - Have a working knowledge of our various test infrastructure tools, such as Testgrid, Triage, Spyglass, Prow, and Tide.

Reviewer 2: links missing

Author: Added, done.

- Have a working knowledge of GitHub labelling and labels used by the project. Release Signal leads must be able to identify

>   relevant issues and PRs for a given milestone.

Reviewer 2: Which labels should we look for? sounds like something which is also good to know about as shadow.

Author: Added, done.

> - Understand what defines a Release Blocking issue or PR, or know who to contact to determine that information.

Reviewer 2: do we have a some doc about this?

Reviewer 2: maybe in the release lead handbook

Author: Are you referring to the Release Signal Lead Checklist? I don't understand.

- Release Signal lead needs to understand what tests matter and how our testing infra is wired together.
    - They can ask previous Release Signal leads for advice.
    - They can ask SIG-Testing for guidance.
- The Release Signal lead should take
  the [Inclusive Speaker Orientation (LFC101)](https://training.linuxfoundation.org/training/inclusive-speaker-orientation/)
  training course.

## Task Overview

> Since the role was created by the merger of the CI Signal sub-team and the Bug Triage sub-team, two project boards must be
> maintained. Detailed information on project boards is

Reviewer 2: ```suggestion
Since the role was created by the merger of the CI Signal sub-team and the Bug Triage sub-team, two project boards must be
maintained. Perhaps this can be unified down the road. Detailed information on project boards is
```

Reviewer 2: maybe? idk, what do you think

Author: I see the possibility, but with all the scrips and such a different schema, it would be a bit challenging. But it would be awesome to have just one reference board instead of two.

Added, done.

       see [this](https://kubernetes.slack.com/archives/CN0K3TE2C/p1694239448798339) thread in
       the [#release-ci-signal](https://kubernetes.slack.com/archives/CN0K3TE2C) channel.

- For issues and PRs targeted for the current release cycle:
    1. Leave a comment on the GitHub issue or PR: "This issue hasn't been updated in 3 months. Are you still expecting to complete
       it for 1.28?". It's helpful to @ mention individuals or SIG ```-bugs``` or ```-pr-reviews``` alias, e.g.
       "@kubernetes/sig-node-bugs" or
       "@kubernetes/sig-network-pr-reviews".
    2. Send a message to the relevant SIG Slack channel or mailing list about the problem: It's helpful to directly @ mention the
       relevant SIG Leads / Owners and to condense multiple issues into a list, e.g. "Hey, these three issues haven't seen any
       attention, are they still valid for 1.28?"
    3. Message individual owners and reviewers directly via Slack. If owners and reviewers are unresponsive, you can reach out to
       the SIG leads using the `kubernetes-sig-${name}-leads@googlegroups.com` mailing lists.

> If none of the methods mentioned above work, you can escalate to the Release Team Lead with suggestions on what to do with
> non-responsive issues.

Reviewer 2: ```suggestion
If none of the methods mentioned above work, you can escalate to the EA or Release Team Leads with suggestions on what to do with
non-responsive issues.
```

EA, lead shadows, lead

Author: Done.

  which grants access to the [ci signal project board](https://github.com/orgs/kubernetes/projects/68/). _Coordinate with the
  release lead to make multiple changes to this file in one PR._
- Adding the lead and (more experienced) shadows as milestone maintainers in
  the [teams.yaml](https://github.com/kubernetes/org/blob/master/config/kubernetes/sig-release/teams.yaml).
  _Coordinate with the release lead to make multiple changes to this file in one PR._
- Plan release support and status reporting.
  See [CI Signal GitHub Projects Board, view: 1.XX-reporting](https://github.com/orgs/kubernetes/projects/68). If the view of the
  current version does not yet exist, a view of a previous version can be duplicated and the filters adjusted.
- Release Signal Shadows must be a member of the Kubernetes organization and, therefore, apply for membership by opening an issue
  on [kubernetes/org](https://github.com/kubernetes/org) (
  see [Issue template](https://github.com/kubernetes/org/issues/new?assignees=&labels=area%2Fgithub-membership&template=membership.yml&title=REQUEST%3A+New+membership+for+%3Cyour-GH-handle%3E)).

> - Organizing an onboarding meeting with shadows to walk through this handbook and useful tools like TestGrid, Spyglass, and

Reviewer 2: links missing

Author: Added, done.

  relevant issues and PRs for a given milestone.
- Understand what defines a Release Blocking issue or PR, or know who to contact to determine that information.
- Release Signal lead needs to understand what tests matter and how our testing infra is wired together.
    - They can ask previous Release Signal leads for advice.
    - They can ask SIG-Testing for guidance.
- The Release Signal lead should take
  the [Inclusive Speaker Orientation (LFC101)](https://training.linuxfoundation.org/training/inclusive-speaker-orientation/)
  training course.

The core responsibility of the Release Signal lead is to foster a culture of continuous test integration, fail-fast and fix-fast
mindset and strive for continuously passing tests that provide an accurate positive signal to the release team to ensure that
releases go out on time with high quality. It is critical to maintain a sustained focus on test health through the entire release
cycle, as opposed to accepting destabilizing test failures during active enhancement development followed by an extended test
stabilization phase.

## Timeline

### Onboarding

- Update

>   the [ci signal team in the teams.yaml](https://github.com/kubernetes/org/blob/main/config/kubernetes/sig-release/teams.yaml)

Reviewer: `[ci signal team in the teams.yaml] `
Shouldn't this be changed to `[release-signal team in the teams.yaml]`?

Author: Not really sure about the changes that need to be done on the release engineering side. Could be like mapping to some structs/GitHub teams and a lot of other things.

@Priyankasaggu11929 What about the changes on the release engineering side? Can we enumerate them so we don't miss out on anything. And a reminder to update the handbook once those changes are in place?

Reviewer 3: >What about the changes on the release engineering side? Can we enumerate them so we don't miss out on anything. And a reminder to update the handbook once those changes are in place?

By release-engineering ^, are we talking about setting up GitHub teams, Slack, and mailing list access/permissions for the lead and shadows in the new "Release Signal" team?

If yes, for  v1.30 and beyond, how about we rename `ci-signal` teams to `release-signal` and remove `bug-triage` teams? Assuming CI signal has more privileges than Bug Triage, this should cover our bases on access and permissions, right?

Reviewer 3: Also, I'm ok leaving it as a TODO with a note (to follow up and update) in the handbook document.

Though, in the TODO note, let's put a link to this particular thread of the PR for context.

Author: > By release-engineering ^, are we talking about

Yes

> If yes, for v1.30 and beyond, how about we rename ci-signal teams to release-signal and remove bug-triage teams? Assuming CI signal has more privileges than Bug Triage, this should cover our bases on access and permissions, right?

I don't think it works that way. For eg- Only folks in the @kubernetes/release-team-bug-triage can edit the bug triage board. We would now be managing two boards (GH project) under one team, so changes would have to be made accordingly. Plus if there are some other permissions granted to bug triage members.

Author: > Also, I'm ok leaving it as a TODO with a note (to follow up and update during v1.30 cycle) in the handbook document.

Sounds good. I'll add a TODO with a link to this review.

Author: Done, added TODO comment

Reviewer 3: >I don't think it works that way. For eg- Only folks in the @kubernetes/release-team-bug-triage can edit the bug triage board. We would now be managing two boards (GH project) under one team, so changes would have to be made accordingly. 

few questions:

- there'll be a single GH board for the new integrated role. Is that correct? 

> Plus if there are some other permissions granted to bug triage members.

No other permissions granted in configs – 
https://cs.k8s.io/?q=release-team-bug-triage&i=nope&files=&excludeFiles=&repos=

I am seeing two usecases only:
1. we are using the ` @kubernetes/release-team-bug-triage` github team to  bulk add folks to the respective GH board
2. or to reference current release-tram bug triage members in github issues/prs/etc.

@hailkomputer, are we using ` kubernetes/release-team-bug-triage` github team for anything else?

Author: > there'll be a single GH board for the new integrated role. Is that correct?

No, the schema is different and merging them will require much work. https://github.com/kubernetes/sig-release/blob/master/release-team/hack/sync-bug-triage-github-project-beta.sh would be needed to be updated as well. So having separate boards is the best option now.

> I am seeing two usecases only:

I am not sure, but maybe for both.

Author: > I am leaning towards all shadows having same access, so, all can fill in wherever required.
Therefore consolidating current release-team-bug-triage and ci-signal github teams into a new release-signal github team and adding all shadows there

Exactly. We require consolidation and not just simple removal.

Reviewer 3: > We require consolidation and not just simple removal.

So, if we agree to this implementation (and maybe want to document it now), then it would mean the following change in [this section of teams.yaml](https://github.com/kubernetes/org/blob/51b4b9c4247bd72c4660c4fff9211b4d3f1b54f2/config/kubernetes/sig-release/teams.yaml#L327)  during start of v1.30 cycle

(also, github and all infra access/perms change are done by Release Lead, so, ok to leave details in role handbook):

- remove existing `ci-signal` and `release-team-bug-triage` teams.
- add a new team `release-team-release-signal` team and add members under it.

The new team would look like:

```
        teams:
          release-team-release-signal:
            description: Members of the Release Signal team for the current release
              cycle.
            members:
             - # 1.30 Release Signal Lead
            -  # 1.30 Release Signal Shadow
            -  # 1.30 Release Signal Shadow            - 
            -  # 1.30 Release Signal Shadow
            privacy: closed
```

Peribolos will then sync the config in the actual GitHub org.

Reviewer 4: > I am seeing two usecases only:
> 
> 1. we are using the ` @kubernetes/release-team-bug-triage` github team to  bulk add folks to the respective GH board
> 2. or to reference current release-tram bug triage members in github issues/prs/etc.

@Vyom-Yadav Correct, those are the only use cases for the group.

Author: @Priyankasaggu11929 I can't find where the team name gives access to the project board. If it's a name-wise mapping, then we may have to update it as now `release-signal` would provide access to `ci-signal` and `bug-triage.` 

I updated the onboarding guide and removed the TODO comment.

Reviewer 3: > I can't find where the team name gives access to the project board. If it's a name-wise mapping, then we may have to update it as now release-signal would provide access to ci-signal and bug-triage.

No, there's no mapping configured anywhere. It's just the github teams can be used to bulk provide access/permissions to all members of a role in a GitHub oroject board (like currently done in case of ci-signal board)
![image](https://github.com/kubernetes/sig-release/assets/30499743/3a778c27-d24f-4485-bdfb-7e4b9cf63d2d)

I think we're good with the current documentation for new `release-signal` github team. No more open items from me.

  which grants access to the [ci signal project board](https://github.com/orgs/kubernetes/projects/68/). _Coordinate with the

>   release lead to make multiple changes to this file in one PR._

Reviewer: > which grants access to the [ci signal project board]

And also to the Bug Triage board?

Author: Nice find. Though it is mentioned in `projects-board.md`, it would be better to have onboarding steps for the bug triage board in here too.

@hailkomputer Can you please confirm what onboarding actions you make to grant permissions? Like adding shadows to milestone maintainers group, etc. Couldn't find it in docs.

Reviewer 4: @Vyom-Yadav it was decided in the previous release not to add bug triage shadows to milestone maintainers group because it is usually not needed. Only onboarding performed is adding the members to the @kubernetes/release-team-bug-triage group for project board access.

Author: Thanks for confirming @hailkomputer. I updated the onboarding guide.

- Adding the lead and (more experienced) shadows as milestone maintainers in
  the [teams.yaml](https://github.com/kubernetes/org/blob/master/config/kubernetes/sig-release/teams.yaml).
  _Coordinate with the release lead to make multiple changes to this file in one PR._
- Plan release support and status reporting.

>   See [CI Signal GitHub Projects Board, view: 1.XX-reporting](https://github.com/orgs/kubernetes/projects/68). If the view of the

Reviewer: The same here: how about Bug Triage board?

Author: https://github.com/kubernetes/sig-release/pull/2352#discussion_r1397659054

Author: Done.

green/yellow/red signals can be initially checked from the table below (Note: Release cutting has only Go/No-Go, so yellow is
equivalent to red).

See [Blocking vs. Informing Dashboards](#blocking-vs-informing-dashboards) for the reason behind these signals.

| (master/release-x.y)-blocking | (master/release-x.y)-informing | Signal |
|-------------------------------|--------------------------------|--------|
| ✅                             | ✅                              | 🟢     |
| ✅                             | ⚠️                             | 🟢     |
| ✅                             | ❌                              | 🔴     |
| ⚠️                            | ⚠️                             | 🟡     |
| ⚠️                            | ✅                              | 🟡     |
| ⚠️                            | ❌                              | 🔴     |
| ❌                             | ⚠️                             | 🔴     |
| ❌                             | ✅                              | 🔴     |
| ❌                             | ❌                              | 🔴     |

Key:

- ❌ - Failing Test

> - ⚠️ - Flaking Test

Reviewer 5: Some flakings are at a low rate and some are at a high rate. 

Should we give some general suggestions about it?

If a flake happens every day and rates more than 2 or 3 in the recent 10 runs, it is yellow for master blocking.
But if the rate is <= 1/10, it could be green.

Author: I am inclining more towards not including that. The reason is that a flake can increase its failure rate. A good example can be observed in this release only. 

https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-reboot

We had four failures recently; before that, it failed infrequently. This ultimately gives every flake equal status so we do not become casual about a flake.

### Closing Issues

After a fix is applied to an issue, it often gets fast closed. Be aware that the issue's status (
open/closed) is not the same as it is on the CI Signal board (New/under investigation/observing/resolved). For the same reasons,
as explained earlier, we have to see if a flaky test got stable. When a fix is applied and it is expected, that this should solve
the problem, the issue gets moved to Observing. After an appropriate amount of time, a big thumb rule would be two weeks, the
issue can be moved to resolved. An issue should remain open until the appropriate fix has been made and the affected tests have
returned to a healthy status. If an issue is closed prematurely, or the same test starts failing or flaking again, a new issue
should be opened. Do not reopen closed issues.

## Special high risk test categories to monitor

Historically, there are a couple of families of tests that are hard to stabilize, regression-prone prone and pose a high risk of
delaying and/or derailing a release. As Release Signal lead, it is highly recommended to pay close attention and extra caution
when dealing with test failures in the following areas.

### Scalability tests

> Scalability testing is inherently challenging, and regressions in this area are potentially a huge project risk

Author: I have left the scalability tests part in, but I am unsure whether this is still required. From 1.27, we didn't do anything special for scalability tests. Any failures involving any SIG have been treated at the same level, i.e., to resolve the issue ASAP.

Points are valid, no problem keeping this in, but it just adds to the length of the handbook as we treat all issues with utmost urgency.
